File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/polarity/pathway_router.py
from __future__ import annotations

from .rule2_validate import find_same_target_polarity_contradictions, validate_rule2_record


def _empty_routing(
    slot_id: str,
    target_id: str | None,
    pathway: str,
    potency_status: str,
    reason_codes: list[str],
    limitation_codes: list[str],
) -> dict:
    return {
        "formula_slot_id": slot_id,
        "formula_target_id": target_id,
        "rule2_record_id": None,
        "disease_polarity": None,
        "required_therapeutic_polarity": None,
        "resolution_status": None,
        "pathway": pathway,
        "potency_status": potency_status,
        "selected_cascade": None,
        "selected_dilution": None,
        "reason_codes": sorted(set(reason_codes)),
        "limitation_codes": sorted(set(limitation_codes)),
    }


def _route_validated_record(record: dict, contradictory_targets: set[str]) -> dict:
    base_limitation = ["PHASE4_NO_NUMERIC_CASCADE"]
    target = record["formula_target_id"]
    if target in contradictory_targets:
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": record["disease_polarity"],
            "required_therapeutic_polarity": record["required_therapeutic_polarity"],
            "resolution_status": record["resolution_status"],
            "pathway": "POLARITY_CONTRADICTORY",
            "potency_status": "UNRESOLVED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": ["POLARITY_SAME_TARGET_CONTRADICTION"],
            "limitation_codes": base_limitation,
        }

    disease = record["disease_polarity"]
    therapeutic = record["required_therapeutic_polarity"]
    status = record["resolution_status"]

    if status in {"CONTRADICTORY", "AMBIGUOUS"}:
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "POLARITY_CONTRADICTORY"
            if status == "CONTRADICTORY"
            else "UNRESOLVED_NO_CASCADE",
            "potency_status": "UNRESOLVED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": [
                "POLARITY_SAME_TARGET_CONTRADICTION"
                if status == "CONTRADICTORY"
                else "RULE2_POLARITY_STATUS_AMBIGUOUS"
            ],
            "limitation_codes": base_limitation,
        }

    if disease == "SUPPORT_ONLY":
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "SUPPORT_ONLY_NON_POTENCY",
            "potency_status": "UNRESOLVED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": ["POTENCY_NOT_APPLICABLE_FOR_SUPPORT_ONLY_SLOT"],
            "limitation_codes": base_limitation,
        }

    if (
        disease == "UNRESOLVED"
        or status == "UNRESOLVED"
        or status == "NEUTRAL_FALLBACK_PENDING_REVIEW"
    ):
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "UNRESOLVED_NO_CASCADE",
            "potency_status": "UNRESOLVED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": ["UPSTREAM_TARGET_POLARITY_NOT_RESOLVED"],
            "limitation_codes": base_limitation,
        }

    if disease == "NEUTRAL" and status == "RESOLVED":
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "NEUTRAL_NON_POTENCY",
            "potency_status": "NOT_EVALUATED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": ["POTENCY_NOT_APPLICABLE_FOR_NEUTRAL_TARGET"],
            "limitation_codes": base_limitation,
        }

    if disease == "MIXED":
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "UNRESOLVED_NO_CASCADE",
            "potency_status": "UNRESOLVED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": [
                "UPSTREAM_TARGET_POLARITY_NOT_RESOLVED",
                "REGISTRY_Q9_SELECTOR_BLOCKED",
            ],
            "limitation_codes": base_limitation,
        }

    if disease == "NEGATIVE" and therapeutic == "POSITIVE" and status == "RESOLVED":
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "NEGATIVE_DISEASE_POSITIVE_THERAPEUTIC_GROUP",
            "potency_status": "NOT_EVALUATED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": ["RULE4_POLARITY_PATHWAY_ROUTED"],
            "limitation_codes": base_limitation,
        }

    if disease == "POSITIVE" and therapeutic == "NEGATIVE" and status == "RESOLVED":
        return {
            "formula_slot_id": record["formula_slot_id"],
            "formula_target_id": target,
            "rule2_record_id": record["rule2_record_id"],
            "disease_polarity": disease,
            "required_therapeutic_polarity": therapeutic,
            "resolution_status": status,
            "pathway": "POSITIVE_DISEASE_NEGATIVE_THERAPEUTIC_GROUP",
            "potency_status": "NOT_EVALUATED",
            "selected_cascade": None,
            "selected_dilution": None,
            "reason_codes": ["RULE4_POLARITY_PATHWAY_ROUTED"],
            "limitation_codes": base_limitation,
        }

    return {
        "formula_slot_id": record["formula_slot_id"],
        "formula_target_id": target,
        "rule2_record_id": record["rule2_record_id"],
        "disease_polarity": disease,
        "required_therapeutic_polarity": therapeutic,
        "resolution_status": status,
        "pathway": "POLARITY_CONTRADICTORY",
        "potency_status": "UNRESOLVED",
        "selected_cascade": None,
        "selected_dilution": None,
        "reason_codes": ["RULE2_THERAPEUTIC_POLARITY_MISMATCH"],
        "limitation_codes": base_limitation,
    }


def route_polarity_for_slots(formula_slot_ids: list[str], records: list[dict]) -> list[dict]:
    contradictory_targets = find_same_target_polarity_contradictions(records)
    by_slot = {r["formula_slot_id"]: r for r in records}
    routings: list[dict] = []
    for slot_id in sorted(formula_slot_ids):
        record = by_slot.get(slot_id)
        if not record:
            routings.append(
                _empty_routing(
                    slot_id,
                    None,
                    "UNRESOLVED_NO_CASCADE",
                    "UNRESOLVED",
                    ["RULE2_POLARITY_RECORD_MISSING"],
                    ["PHASE4_NO_NUMERIC_CASCADE"],
                )
            )
            continue
        validated = validate_rule2_record(record)
        if not validated["ok"]:
            routings.append(
                _empty_routing(
                    slot_id,
                    record.get("formula_target_id"),
                    "UNRESOLVED_NO_CASCADE",
                    "UNRESOLVED",
                    validated["reason_codes"],
                    ["PHASE4_NO_NUMERIC_CASCADE"],
                )
            )
            continue
        routings.append(_route_validated_record(record, contradictory_targets))
    return sorted(routings, key=lambda r: r["formula_slot_id"])

File: packages/clinical-contracts/src/rule4/polarity/pathwayRouter.ts
import type {
  Rule2FormulaPolarityRecord,
  Rule4PolarityPathway,
  Rule4SlotPolarityRouting,
} from './types.js';
import { findSameTargetPolarityContradictions, validateRule2Record } from './rule2Validate.js';

function emptyRouting(
  slotId: string,
  targetId: string | null,
  pathway: Rule4PolarityPathway,
  potencyStatus: Rule4SlotPolarityRouting['potencyStatus'],
  reasonCodes: string[],
  limitationCodes: string[],
): Rule4SlotPolarityRouting {
  return {
    formulaSlotId: slotId,
    formulaTargetId: targetId,
    rule2RecordId: null,
    diseasePolarity: null,
    requiredTherapeuticPolarity: null,
    resolutionStatus: null,
    pathway,
    potencyStatus,
    selectedCascade: null,
    selectedDilution: null,
    reasonCodes: [...new Set(reasonCodes)].sort(),
    limitationCodes: [...new Set(limitationCodes)].sort(),
  };
}

function routeValidatedRecord(
  record: Rule2FormulaPolarityRecord,
  contradictoryTargets: Set<string>,
): Rule4SlotPolarityRouting {
  const baseLimitation = ['PHASE4_NO_NUMERIC_CASCADE'];
  if (contradictoryTargets.has(record.formulaTargetId)) {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: record.diseasePolarity,
      requiredTherapeuticPolarity: record.requiredTherapeuticPolarity,
      resolutionStatus: record.resolutionStatus,
      pathway: 'POLARITY_CONTRADICTORY',
      potencyStatus: 'UNRESOLVED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['POLARITY_SAME_TARGET_CONTRADICTION'],
      limitationCodes: baseLimitation,
    };
  }

  const disease = record.diseasePolarity;
  const therapeutic = record.requiredTherapeuticPolarity;
  const status = record.resolutionStatus;

  if (status === 'CONTRADICTORY' || status === 'AMBIGUOUS') {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: status === 'CONTRADICTORY' ? 'POLARITY_CONTRADICTORY' : 'UNRESOLVED_NO_CASCADE',
      potencyStatus: 'UNRESOLVED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: [
        status === 'CONTRADICTORY'
          ? 'POLARITY_SAME_TARGET_CONTRADICTION'
          : 'RULE2_POLARITY_STATUS_AMBIGUOUS',
      ],
      limitationCodes: baseLimitation,
    };
  }

  if (disease === 'SUPPORT_ONLY') {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: 'SUPPORT_ONLY_NON_POTENCY',
      potencyStatus: 'UNRESOLVED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['POTENCY_NOT_APPLICABLE_FOR_SUPPORT_ONLY_SLOT'],
      limitationCodes: baseLimitation,
    };
  }

  if (
    disease === 'UNRESOLVED' ||
    status === 'UNRESOLVED' ||
    status === 'NEUTRAL_FALLBACK_PENDING_REVIEW'
  ) {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: 'UNRESOLVED_NO_CASCADE',
      potencyStatus: 'UNRESOLVED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['UPSTREAM_TARGET_POLARITY_NOT_RESOLVED'],
      limitationCodes: baseLimitation,
    };
  }

  if (disease === 'NEUTRAL' && status === 'RESOLVED') {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: 'NEUTRAL_NON_POTENCY',
      potencyStatus: 'NOT_EVALUATED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['POTENCY_NOT_APPLICABLE_FOR_NEUTRAL_TARGET'],
      limitationCodes: baseLimitation,
    };
  }

  if (disease === 'MIXED') {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: 'UNRESOLVED_NO_CASCADE',
      potencyStatus: 'UNRESOLVED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['UPSTREAM_TARGET_POLARITY_NOT_RESOLVED', 'REGISTRY_Q9_SELECTOR_BLOCKED'],
      limitationCodes: baseLimitation,
    };
  }

  if (disease === 'NEGATIVE' && therapeutic === 'POSITIVE' && status === 'RESOLVED') {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: 'NEGATIVE_DISEASE_POSITIVE_THERAPEUTIC_GROUP',
      potencyStatus: 'NOT_EVALUATED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['RULE4_POLARITY_PATHWAY_ROUTED'],
      limitationCodes: baseLimitation,
    };
  }

  if (disease === 'POSITIVE' && therapeutic === 'NEGATIVE' && status === 'RESOLVED') {
    return {
      formulaSlotId: record.formulaSlotId,
      formulaTargetId: record.formulaTargetId,
      rule2RecordId: record.rule2RecordId,
      diseasePolarity: disease,
      requiredTherapeuticPolarity: therapeutic,
      resolutionStatus: status,
      pathway: 'POSITIVE_DISEASE_NEGATIVE_THERAPEUTIC_GROUP',
      potencyStatus: 'NOT_EVALUATED',
      selectedCascade: null,
      selectedDilution: null,
      reasonCodes: ['RULE4_POLARITY_PATHWAY_ROUTED'],
      limitationCodes: baseLimitation,
    };
  }

  return {
    formulaSlotId: record.formulaSlotId,
    formulaTargetId: record.formulaTargetId,
    rule2RecordId: record.rule2RecordId,
    diseasePolarity: disease,
    requiredTherapeuticPolarity: therapeutic,
    resolutionStatus: status,
    pathway: 'POLARITY_CONTRADICTORY',
    potencyStatus: 'UNRESOLVED',
    selectedCascade: null,
    selectedDilution: null,
    reasonCodes: ['RULE2_THERAPEUTIC_POLARITY_MISMATCH'],
    limitationCodes: baseLimitation,
  };
}

export function routePolarityForSlots(
  formulaSlotIds: readonly string[],
  records: readonly Rule2FormulaPolarityRecord[],
): Rule4SlotPolarityRouting[] {
  const contradictoryTargets = findSameTargetPolarityContradictions(records);
  const bySlot = new Map<string, Rule2FormulaPolarityRecord>();
  for (const r of records) {
    bySlot.set(r.formulaSlotId, r);
  }

  const routings: Rule4SlotPolarityRouting[] = [];
  for (const slotId of [...formulaSlotIds].sort()) {
    const record = bySlot.get(slotId);
    if (!record) {
      routings.push(
        emptyRouting(
          slotId,
          null,
          'UNRESOLVED_NO_CASCADE',
          'UNRESOLVED',
          ['RULE2_POLARITY_RECORD_MISSING'],
          ['PHASE4_NO_NUMERIC_CASCADE'],
        ),
      );
      continue;
    }
    const validated = validateRule2Record(record);
    if (!validated.ok) {
      routings.push(
        emptyRouting(
          slotId,
          record.formulaTargetId ?? null,
          'UNRESOLVED_NO_CASCADE',
          'UNRESOLVED',
          validated.reasonCodes,
          ['PHASE4_NO_NUMERIC_CASCADE'],
        ),
      );
      continue;
    }
    routings.push(routeValidatedRecord(record, contradictoryTargets));
  }

  return routings.sort((a, b) => a.formulaSlotId.localeCompare(b.formulaSlotId));
}
